prov/rxm/src/rxm_conn.c

```diff
@@ -1089,14 +1093,39 @@ static int rxm_conn_handle_notify(struct fi_eq_entry *eq_entry)
 		assert(handle->state == RXM_CMAP_SHUTDOWN);
 		FI_DBG(&rxm_prov, FI_LOG_EP_CTRL, "freeing handle: %p\n", handle);
 		cmap = handle->cmap;
+		rxm_ep = container_of(cmap->ep, struct rxm_ep, util_ep);
+
+		rxm_conn_free(handle);
+
+		// after closing the connection, we need to flush any dangling references to the
+		// handle from msg_cq entries that have not been cleaned up yet, otherwise we
+		// could run into problems during CQ cleanup.  these entries will be errored so
+		// keep reading through EAVAIL.
+		do {
+			ret = fi_cq_read(rxm_ep->msg_cq, &comp, 1);
+			if (ret > 0) {
+				ret = rxm_cq_handle_comp(rxm_ep, &comp);
+				if (OFI_UNLIKELY(ret)) {
+					rxm_cq_write_error_all(rxm_ep, ret);
+				} else {
+					ret = 1;
+				}
+			} else if (ret == -FI_EAVAIL) {
+				rxm_cq_read_write_error(rxm_ep);
+				ret = 1;
+			} else if (ret < 0 && ret != -FI_EAGAIN) {
+				rxm_cq_write_error_all(rxm_ep, ret);
+			}
+		} while (ret > 0);
```

Comment:
I would move this into a function (e.g. rxm_flush_msg_cq) to improve readability.